### ISSUE TITLE
Investigate and fix Training test group failures

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#2531]
## Problem

The "Training" test group in PR #2528 is failing, but the root cause has not been identified yet because CI logs were not accessible during initial investigation.

## Investigation Required

The Training test group runs all `test_*.mojo` files in `tests/shared/training/` which includes 26 test files:
- `test_optimizers.mojo` - SGD, Adam, AdamW, RMSprop tests
- `test_schedulers.mojo` - CosineAnnealingLR, ReduceLROnPlateau tests
- `test_training_loop.mojo` - Training loop tests (some TDD placeholders)
- `test_callbacks.mojo` - Training callbacks
- `test_checkpointing.mojo` - Checkpoint save/load
- `test_metrics.mojo` - Training metrics tracking
- `test_evaluation.mojo` - Model evaluation
- And 19 more test files

## Possible Failure Scenarios

1. **TDD Placeholder Tests** (High probability)
   - Some tests like `test_training_loop.mojo` have placeholder implementations
   - Tests documented with `# TODO(#34): Implement when TrainingLoop is available`
   
2. **Import/Compilation Errors** (Medium probability)
   - Missing module exports in `__init__.mojo` files
   - Trait conformance issues (LRScheduler trait)
   - Cross-module import failures
   
3. **Runtime Errors** (Lower probability)
   - Type mismatches in function signatures
   - Numerical assertion failures
   - Memory access violations

## Investigation Steps

1. **Fetch actual CI failure logs**:
   ```bash
   gh run view 20049407984 --log-failed > /tmp/training-failure.log
   grep -B 10 -A 100 "Training" /tmp/training-failure.log
   ```

2. **Identify error type**:
   ```bash
   # Check for compilation errors
   grep -E "error:|failed to compile" /tmp/training-failure.log
   
   # Check for import errors
   grep -E "cannot find|module not found" /tmp/training-failure.log
   
   # Check for runtime errors
   grep -E "FAILED|panic:|AssertionError" /tmp/training-failure.log
   ```

3. **Locate failing test file**:
   ```bash
   grep -B 5 "error:" /tmp/training-failure.log | grep "test_.*\.mojo"
   ```

## Files to Check

**High Priority** (known to have complex implementations):
- `tests/shared/training/test_warmup_scheduler.mojo` - Full implementation
- `tests/shared/training/test_optimizers.mojo` - May have import issues
- `tests/shared/training/test_schedulers.mojo` - May have import issues

**Known Placeholders**:
- `tests/shared/training/test_training_loop.mojo` - Most tests commented out (Issue #34)
- `tests/shared/training/test_validation_loop.mojo` - Documented as skipped (Issue #34)

## Dependencies

Training tests import from:
- `shared.training.optimizers.*` - SGD, Adam, AdamW, RMSprop
- `shared.training.schedulers.*` - CosineAnnealingLR, ReduceLROnPlateau, WarmupLR
- `shared.training.base.*` - LRScheduler trait, gradient utilities
- `shared.core.*` - ExTensor, tensor operations

## Success Criteria

- [ ] Identify exact failing test file(s)
- [ ] Categorize failure type (compilation/runtime/TDD)
- [ ] Fix compilation/import errors if found
- [ ] Document TDD tests as expected failures
- [ ] All Training tests either pass or are documented as pending

## Related

- PR #2528 - Fix build issues
- Issue #34 - Referenced in training loop placeholders
- CI Run: https://github.com/mvillmow/ml-odyssey/actions/runs/20049407984
